compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs

```diff
@@ -1577,33 +1577,28 @@ impl<'tcx> InferCtxtPrivExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
         }
 
         self.probe(|_| {
-            let mut err = error.err;
-            let mut values = None;
+            let ocx = ObligationCtxt::new_in_snapshot(self);
 
             // try to find the mismatched types to report the error with.
             //
             // this can fail if the problem was higher-ranked, in which
             // cause I have no idea for a good error message.
             let bound_predicate = predicate.kind();
-            if let ty::PredicateKind::Clause(ty::Clause::Projection(data)) =
+            let (values, err) = if let ty::PredicateKind::Clause(ty::Clause::Projection(data)) =
                 bound_predicate.skip_binder()
             {
-                let mut selcx = SelectionContext::new(self);
                 let data = self.replace_bound_vars_with_fresh_vars(
                     obligation.cause.span,
                     infer::LateBoundRegionConversionTime::HigherRankedType,
                     bound_predicate.rebind(data),
                 );
-                let mut obligations = vec![];
-                // FIXME(normalization): Change this to use `At::normalize`
-                let normalized_ty = super::normalize_projection_type(
-                    &mut selcx,
-                    obligation.param_env,
-                    data.projection_ty,
-                    obligation.cause.clone(),
-                    0,
-                    &mut obligations,
-                );
+                let InferOk { value: normalized_ty, .. } =
+                    self.at(&obligation.cause, obligation.param_env).normalize(
```

Comment:
so `ocx.normalize` here actually

Reply:
my suggestion was wrong if we're gonna call `select_obligations_*` below

Reply:
this should still use `ocx.normalize`

Reply:
@lcnr can you explain why `ocx.normalize` and not `self.at(..).normalize`?

Reply:
Because `ocx.normalize` just calls `infcx.at.normalize` and registers the obligations -- First, you're throwing away the obligations, and second, if we're gonna be using other `ocx` methods, then it's simpler.

Reply:
I was confused by the existing comment, anyway, done.